pkg/controller/clone-controller.go

```diff
@@ -102,162 +185,150 @@ func (c *CloneController) Run(threadiness int, stopCh <-chan struct{}) error {
 // processing this pvc, which would result in multiple pods for the same pvc.
 func (c *CloneController) processPvcItem(pvc *v1.PersistentVolumeClaim) error {
-	cr, err := getCloneRequestPVC(pvc)
+	anno := map[string]string{}
+	
+	// find import Pod
+	pods, err := c.findClonePodsFromCache(pvc)
 	if err != nil {
 		return err
 	}
-
-	// check our existing pvc one more time to ensure we should be working on it
-	// and to help mitigate any race conditions. This time we get the latest pvc.
-	doCreate, pvc, err := checkClonePVC(c.clientset, pvc, true)
-	if err != nil { // maybe an intermittent api error
+	pvcKey, err := cache.MetaNamespaceKeyFunc(pvc)
+	if err != nil {
 		return err
 	}
-	if !doCreate { // don't create pods but not an error
-		return nil // forget key; logging already done
-	}
 
-	// update pvc with 'AnnCloningPods' annotation to indicate cloning is in process
-	anno := map[string]string{AnnCloningPods: "exist"}
-	var lab map[string]string
-	pvc, err = updatePVC(c.clientset, pvc, anno, lab)
-	if err != nil {
-		return errors.WithMessage(err, "could not update pvc %q annotation and/or label")
+	// Pods must be controlled by this PVC
+	for _, pod := range pods {
+		if pod != nil && !metav1.IsControlledBy(pod, pvc) {
+			return errors.Errorf("found pod %s/%s not owned by pvc %s/%s", pod.Namespace, pod.Name, pvc.Namespace, pvc.Name)
+	 	}
 	}
 
-	//create random string to be used for pod labeling and hostpath name
-	generatedLabelStr := util.RandAlphaNum(GENERATED_CLONING_LABEL_LEN)
-	//create the source pod
-	pod, err := CreateCloneSourcePod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, cr, pvc, generatedLabelStr)
-	if err != nil {
-		//TODO: remove annotation AnnCloningPods from pvc as pod failed to run
-		return err
+	// expectations prevent us from creating multiple pods. An expectation forces
+	// us to observe a pod's creation in the cache.
+	needsSync := c.podExpectations.SatisfiedExpectations(pvcKey)
+
+	// make sure not to reprocess a PVC that has already completed successfully,
+	// even if the pod no longer exists
+	previousPhase, exists := pvc.ObjectMeta.Annotations[AnnClonePodPhase]
+	if exists && (previousPhase == string(v1.PodSucceeded)) {
+		needsSync = false
 	}
 
-	//create the target pod
-	_, err = CreateCloneTargetPod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, pvc, generatedLabelStr, pod.ObjectMeta.Namespace)
-	if err != nil {
-		//TODO: remove annotation AnnCloningPods from pvc as pod failed to run
-		return err
+	if pods == nil && needsSync {
+		cr, err := getCloneRequestPVC(pvc)
+		if err != nil {
+			return err
+		}
+
+		// all checks passed, let's create the cloner pods!
+		c.expectPodCreate(pvcKey)
+		//create random string to be used for pod labeling and hostpath name
+		generatedLabelStr := util.RandAlphaNum(GENERATED_CLONING_LABEL_LEN)
+		//create the source pod
+		pod, err := CreateCloneSourcePod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, cr, pvc, generatedLabelStr)
```

Comment:
this needs to be idempotent. only create the source pod if it doesn't already exist yet.

Reply:
did not get it. We create the pods only if the returns pods list is nil...

Reply:
> We create the pods only if the returns pods list is nil...

That's not the correct logic for the clone controller. You'll have to check that both the src and target pods are created. 

The ```if pod == nil``` only applies to the importer pod because there is only a single pod associated with the import job. 

So the logic here is something like

```
if needsSync && srcPod ==nil {
// then create the src pod
}
if needsSync && targetPod == nil {
// then create target pod
}
```

The findPods function will have to find both the src and target pods instead of a single pod.

Hopefully that helps!